jgrapht-core/src/main/java/org/jgrapht/alg/tour/ParberryKnightTour.java

```diff
@@ -0,0 +1,412 @@
+         */
+
+        if (tour1.getStructured().get(7).getNext() == tour1.getStructured().get(6)) {
+            tour1.getStructured().get(7).setNext(tour3.getStructured().get(2));
+            tour1.getStructured().get(6).setPrev(tour2.getStructured().get(4));
+        }
+        else {
+            tour1.getStructured().get(7).setPrev(tour3.getStructured().get(2));
+            tour1.getStructured().get(6).setNext(tour2.getStructured().get(4));
+        }
+
+        if (tour3.getStructured().get(2).getPrev() == tour3.getStructured().get(3)) {
+            tour3.getStructured().get(2).setPrev(tour1.getStructured().get(7));
+            tour3.getStructured().get(3).setNext(tour4.getStructured().get(1));
+        }
+        else {
+            tour3.getStructured().get(2).setNext(tour1.getStructured().get(7));
+            tour3.getStructured().get(3).setPrev(tour4.getStructured().get(1));
+        }
+
+        if (tour4.getStructured().get(1).getPrev() == tour4.getStructured().get(0)) {
+            tour4.getStructured().get(1).setPrev(tour3.getStructured().get(3));
+            tour4.getStructured().get(0).setNext(tour2.getStructured().get(5));
+        }
+        else {
+            tour4.getStructured().get(1).setNext(tour3.getStructured().get(3));
+            tour4.getStructured().get(0).setPrev(tour2.getStructured().get(5));
+        }
+
+        if (tour2.getStructured().get(5).getPrev() == tour2.getStructured().get(4)) {
+            tour2.getStructured().get(5).setPrev(tour4.getStructured().get(0));
+            tour2.getStructured().get(4).setNext(tour1.getStructured().get(6));
+        }
+        else {
+            tour2.getStructured().get(5).setNext(tour4.getStructured().get(0));
+            tour2.getStructured().get(4).setPrev(tour1.getStructured().get(6));
+        }
+
+        /*
+         Update the start node after you've contracted all quadrants.
+         */
+
+        tour1.getList().setStartNode(tour3.getStructured().get(2));
+
+        /*
+         Update structured pointers.
+         Note that we do not need to update the first two nodes, since they are already set correctly.
+         */
+
+        tour1.getStructured().set(2, tour2.getStructured().get(2));
+        tour1.getStructured().set(3, tour2.getStructured().get(3));
+
+        tour1.getStructured().set(4, tour3.getStructured().get(4));
+        tour1.getStructured().set(5, tour3.getStructured().get(5));
+
+        tour1.getStructured().set(6, tour4.getStructured().get(6));
+        tour1.getStructured().set(7, tour4.getStructured().get(7));
+
+        /*
+         Update size of the list.
+         */
+
+        tour1.getList().setSize(tour1.getList().getSize() + tour2.getList().getSize() + tour3.getList().getSize() + tour4.getList().getSize());
+
+        return tour1;
+    }
+
+    /**
+     * Returns a closed knight's tour.
+     * @return closed knight's tour.
+     */
+
+    public KnightTour getTour() {
```

Comment:
Can be written shorter: return generateTour(new Pair<>(0, 0), new Pair<>(n - 1, m - 1));